### ISSUE TITLE
How to use webhooks to activate and build new version?

### ISSUE THREAD

**Issue by agoose77** [readthedocs/readthedocs.org#9673]
Hi all, thanks for all of your hard work on RTD, it's an invaluable resource.

I'm trying to improve our docs experience over at https://readthedocs.org/projects/awkward-array/, after moving to a manual webhook-triggered build. Originally we were going to build the entire HTML artefact, and have RTD just serve those contents, but it seems less painful to build the computationally intensive parts in our CI, and have the RTD Sphinx build pull those artefacts down during building.

The need to build first on our CI means that we now use a manual incoming integration hook to trigger the build on RTD. At present, this requires the version being built (e.g. the branch) to be activated on RTD. What I'd like to happen instead is:

- In our CI, tell RTD that to build and activate a new (hidden) version. This happens at the end of our workflow so that the pre-built assets are available. This would allow new PRs to have built documentation.
- Have RTD build our old branch that doesn't use a manual trigger whenever a release / push is made.

Could anyone provide some guidance on this? 

Thanks!

**Comment by humitos** [comment]
Hi @agoose77! Before going too depth into a potential solution to your scenario, I want to ask "Why do you need to build the docs _outside_ Read the Docs and use it only as a hosting? Why you don't build the docs on Read the Docs as well?"

**Comment by agoose77** [comment]
Our docs contain Jupyter Notebooks & use a WASM deployment of our package, both of which take a long time to build. We exhaust the RTD build resources when doing this on RTD.

**Comment by humitos** [comment]
Gotcha! Thanks for the explanation.

> In our CI, tell RTD that to build and activate a new (hidden) version. This happens at the end of our workflow so that the pre-built assets are available. This would allow new PRs to have built documentation.

You could use this API endpoint https://docs.readthedocs.io/en/stable/api/v3.html#version-update to activate a version and this one https://docs.readthedocs.io/en/stable/api/v3.html#build-triggering to trigger a build for that version.

> Have RTD build our old branch that doesn't use a manual trigger whenever a release / push is made.

I'm not 100% sure to understand this one, but I guess that you can use the same API endpoint, https://docs.readthedocs.io/en/stable/api/v3.html#build-triggering, to trigger a build in that particular RTD version.

Let me know if this helps.

**Comment by agoose77** [comment]
Thanks, I saw those yesterday and found myself wondering how to compute the slug for the version, i.e. how does RTD escape special characters for the URL. I imagine that I'll need to do that on our end rather than just providing the raw branch name e.g. `agoose77/fix-something`.

**Comment by agoose77** [comment]
Ah, thanks! I think we can try the simple case first time around, and/or write a small script to invoke unicode slugify.

**Comment by agoose77** [comment]
Is there a better way to achieve our goals than using webhooks here? It feels like we're fighting a lot of the defaults on RTD. For example, I'd like to have PRs build under transient URLs like RTD supports for the GitHub integration. Triggering builds via webhooks is a somewhat orthogonal system.

**Comment by humitos** [comment]
I'm not sure what is your goal and how your build process works exactly. I thought that you only wanted to trigger builds manually after "something else" has completed and for that case, the API looked to work.

You will need to give more details about what exactly you are trying to achieve, but it seems that you want most of the features from Read the Docs but using it in a pretty-custom way. So, I'm not sure that will be possible.

**Comment by agoose77** [comment]
Thanks @humitos. The missing piece is that we want ideally for RTD to wait until our build artefacts are ready before trying to build. If we use the GitHub integration, it usually picks up the push events etc. and immediately starts building. We can put up with this by making the first artefact-free build fail, and then trigger a manual build later. However, this secondary step is not possible (at least, as far as I can tell) for PR-builds; can we trigger PR builds via a version slug?

**Comment by humitos** [comment]
> can we trigger PR builds via a version slug?

I'd say this is not currently supported (cc @stsewd) but -- there should be a way to hack it if you want to invest some time 😄 . I'd try the following:

1. install/re-sync the GitHub integration on the Read the Docs project
1. remove the GitHub webhook from the GH repository that hits Read the Docs to avoid triggering the builds _before_ the artifacts are ready
1. configure the CI that builds the artifacts needed for the docs build to "emulate" [^1] a GitHub request once the artifacts are built, and hit the Read the Docs integration

Of course, this is not officially supported. I'm just trying to find a workaround that could work for your use case. In any case, let me know if this is something you are happy to try.

[^1]: you will need to send a similar request to the one that GitHub sends: https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#pull_request --most of the fields can probably be omitted, since Read the Docs doesn't use them all (_required_ are probably `action`, `created`, `deleted`, the HTTP header `GitHub-Event` and... maybe something else) This is the code that I followed for this: https://github.com/readthedocs/readthedocs.org/blob/a09bc1a976a93bcc3f987fa0a052901f0065619f/readthedocs/api/v2/views/integrations.py#L378

**Comment by agoose77** [comment]
> I'd say this is not currently supported (cc @stsewd) but -- there should be a way to hack it if you want to invest some time smile . I'd try the following:

Right! I had a suspicion that you might suggest this ;) We're definitely in the realms of hacking a solution, so I'll give it a think. In the mean-time, I *think* the current solution that we have: to use the GH webhook with *only* branch events, and using the REST API to trigger builds/activate versions, is sufficient.

Thanks for the help!

**Comment by stsewd** [comment]
> can we trigger PR builds via a version slug?

This isn't possible, but could be supported I guess, but it may require some work, specially to have parity with the current workflow we have (like open and close a PR, send the commit status, etc)
